### PR TITLE
Message retraction: RDBMS

### PR DESCRIPTION
**Pull request by chrzaszcz**
Message retraction support in MAM for RDBMS

This PR introduces the support for [XEP-0424 Message Retraction](https://xmpp.org/extensions/xep-0424.html).
- Retraction messages are stored in MAM archives even though they contain no body, as required by the XEP.
- MAM archives using RDBMS handle the retraction messages by replacing the original message (referenced by `origin-id`) with a tombstone.

TO DO in next pull requests:
- Discovery of the retraction feature.
- Documentation of the retraction feature.
- Possible error replies to the sender when the message to retract from MAM does not exist.

### CODE REVIEW

File: src/mam/ejabberd_gen_mam_archive.erl [hunk 1]
```diff
@@ -7,7 +7,8 @@
 -callback archive_message(_Result, jid:server(),
                           MessID :: mod_mam:message_id(), ArchiveID :: mod_mam:archive_id(),
                           LocJID :: jid:jid(), RemJID :: jid:jid(),
-                          SrcJID :: jid:jid(), Dir :: atom(), Packet :: any()) ->
+                          SrcJID :: jid:jid(), OriginID :: binary(),
+                          Dir :: atom(), Packet :: any()) ->
```
Comment (NelsonVides) [inline src/mam/ejabberd_gen_mam_archive.erl]: Hmm... I remember many discussion with Radek about functions that start taking too many parameters... At some point they become unruly and hard to follow. I'd like to raise here the option to pass a map with the necessary values, it's more readable and extensible, what do you think? 🤔

Comment (chrzaszcz) [reply src/mam/ejabberd_gen_mam_archive.erl]: It's on my list of top thing to do after message retraction is finished. Already discussed with @michalwski and @erszcz.

File: src/mam/mod_mam_rdbms_arch.erl [hunk 2]
```diff
@@ -166,25 +167,70 @@ archive_message(Result, Host, MessID, UserID,
...
     end.
 
-do_archive_message(_Result, Host, MessID, UserID, LocJID, RemJID, SrcJID, Dir, Packet) ->
-    Row = prepare_message(Host, MessID, UserID, LocJID, RemJID, SrcJID, Dir, Packet),
+do_archive_message(_Result, Host, MessID, UserID, LocJID, RemJID, SrcJID, OriginID, Dir, Packet) ->
+    Row = prepare_message(Host, MessID, UserID, LocJID, RemJID, SrcJID, OriginID, Dir, Packet),
     {updated, 1} = mod_mam_utils:success_sql_execute(Host, insert_mam_message, Row),
+    retract_message(Host, MessID, UserID, LocJID, RemJID, SrcJID, OriginID, Dir, Packet).
+
+retract_message(Host, _MessID, UserID, LocJID, RemJID, _SrcJID, _OriginID, Dir, Packet) ->
```
Comment (NelsonVides) [inline src/mam/mod_mam_rdbms_arch.erl:175]: Hmmm... I don't like passing parameters just to ignore them 😕 I see that this is how the async pool writer works though, so I don't really have a suggestion on how to fix both 🤷‍♂️

File: src/mam/mod_mam.erl [hunk 3]
```diff
@@ -589,12 +590,13 @@ lookup_messages_without_policy_violation_check(Host, #{search_text := SearchText
 
 -spec archive_message(Host :: jid:server(), MessID :: message_id(),
                       ArcID :: archive_id(), LocJID :: jid:jid(), RemJID :: jid:jid(),
-                      SrcJID :: jid:jid(), Dir :: incoming | outgoing, Packet :: term()
+                      SrcJID :: jid:jid(), OriginID :: binary(),
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/mam/mod_mam.erl:593]: Good catch! It was `none` before I added the support for MUC, but I forgot to change it.

File: src/mam/mod_mam_muc.erl [hunk 4]
```diff
@@ -493,11 +494,11 @@ lookup_messages_without_policy_violation_check(Host, #{search_text := SearchText
 
 -spec archive_message(jid:server(), MessId :: mod_mam:message_id(),
                       ArcId :: mod_mam:archive_id(), LocJID :: jid:jid(),
-                      SenderJID :: jid:jid(), SrcJID :: jid:jid(), Dir :: 'incoming',
-                      packet()) -> any().
-archive_message(Host, MessID, ArcID, LocJID, SenderJID, SrcJID, Dir, Packet) ->
+                      SenderJID :: jid:jid(), SrcJID :: jid:jid(), OriginID :: none,
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/mam/mod_mam_muc.erl:497]: Good catch! It was `none` before I added the support for MUC, but I forgot to change it.

File: src/mam/mod_mam_rdbms_arch.erl [hunk 5]
```diff
@@ -166,25 +167,70 @@ archive_message(Result, Host, MessID, UserID,
...
+     " WHERE user_id = ", SUserID, " AND id = '", BMessID, "'"].
+
 prepare_message(Host, MessID, UserID, LocJID = #jid{}, RemJID = #jid{lresource = RemLResource},
-                SrcJID, Dir, Packet) ->
+                SrcJID, OriginID, Dir, Packet) ->
     SBareRemJID = jid_to_stored_binary(Host, LocJID, jid:to_bare(RemJID)),
     SSrcJID = jid_to_stored_binary(Host, LocJID, SrcJID),
     SDir = encode_direction(Dir),
+    SOriginID = case OriginID of
+                    none -> null;
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/mam/mod_mam_rdbms_arch.erl:222]: RDBMS needs the atom `null`, other databases might need something else... I am not sure what's best here, I'd rather go with `no_origin_id` than `null`.

File: src/mam/mod_mam_rdbms_arch.erl (hunk 2 above)
Comment (chrzaszcz) [reply src/mam/mod_mam_rdbms_arch.erl:175]: I also don't like it but... it's good news - when we change the arguments to a map (ad I am going to do this soon), we could just pass the map here.

File: src/mam/mod_mam_rdbms_arch.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/mam/mod_mam_rdbms_arch.erl:222]: @NelsonVides just take a look at lines 394-412 of `mod_mam_utils.erl`: https://github.com/esl/MongooseIM/pull/2722/files#diff-700dac120726001b23fa469f54948e36R394-R412

File: src/mam/mod_mam_muc_rdbms_arch.erl [hunk 6]
```diff
@@ -127,12 +127,13 @@ archive_size(Size, Host, RoomID, _RoomJID) when is_integer(Size) ->
 -spec archive_message(_Result, jid:server(), MessID :: mod_mam:message_id(),
                       RoomID :: mod_mam:archive_id(), _LocJID :: jid:jid(),
                       SenderJID :: jid:jid(),
-                      UserRoomJID :: jid:jid(), incoming, Packet :: packet()) -> ok.
+                      UserRoomJID :: jid:jid(), none, incoming, Packet :: packet()) -> ok.
```
Comment (NelsonVides) [inline src/mam/mod_mam_muc_rdbms_arch.erl:130]: Ag, I guess I missed this one, sorry. Spec should be again `binary() | none`. Maybe we could improve indentation already, `SenderJID` has a line for himself and then the next 4 parameters and the return value are all on the next.

File: src/mam/mod_mam_rdbms_arch.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (NelsonVides) [reply src/mam/mod_mam_rdbms_arch.erl:222]: `no_retract_id` and `no_origin_id` are verbose only if we have a clear association between the function that produced the value and the value itself. Like, if we know that this particular `none` came from `get_retract_id`, then clearly `get_retract_id -> no_retract_id` is more verbose than `get_retract_id -> none`. But if we lose track of the value's source then having a bunch of `none`s can be confusing 😕 
In any case from that perspective I had a look at how easy it is to track what is what and seems fine for now, so it's just an observation. The atom that different databases might need though, yeah, maybe that's that 🙂
